addons/controllers/calico/calicoconfig_controller.go

```diff
@@ -6,43 +6,269 @@ package controllers
+		r.Log.Error(err, "Error reconciling CalicoConfig to create data value secret")
+		return ctrl.Result{}, err
+	}
+
+	r.Log.Info("CalicoConfig successfully reconciled")
+	return ctrl.Result{}, nil
+}
+
+// ReconcileCalicoConfigNormal reconciles CalicoConfig by creating data values secret
+func (r *CalicoConfigReconciler) ReconcileCalicoConfigNormal(
+	cluster *clusterapiv1beta1.Cluster,
+	calicoConfig *cniv1alpha1.CalicoConfig,
+	patchCalicoConfig *bool) (retErr error) {
+
+	// add finalizer to CalicoConfig
+	*patchCalicoConfig = util.AddFinalizerToCR(r.Log, constants.CalicoAddonName, calicoConfig)
+
+	// add owner reference to CalicoConfig if not already added by TanzuClusterBootstrap Controller
+	ownerReference := metav1.OwnerReference{
+		APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+		Kind:               cluster.Kind,
+		Name:               cluster.Name,
+		UID:                cluster.UID,
+		Controller:         pointer.BoolPtr(true),
+		BlockOwnerDeletion: pointer.BoolPtr(false),
+	}
+	if !clusterapiutil.HasOwnerRef(calicoConfig.OwnerReferences, ownerReference) {
+		r.Log.Info("Adding owner reference to CalicoConfig")
+		calicoConfig.OwnerReferences = clusterapiutil.EnsureOwnerRef(calicoConfig.OwnerReferences, ownerReference)
+		*patchCalicoConfig = true
+	}
+
+	// create data values secret for CalicoConfig
+	secretNamespacedName := types.NamespacedName{
+		Name:      util.GenerateDataValueSecretName(calicoConfig.Name, constants.CalicoAddonName),
+		Namespace: calicoConfig.Namespace,
+	}
+	if err := r.ReconcileDataValuesSecret(cluster, calicoConfig, secretNamespacedName); err != nil {
+		return err
+	}
+
+	// add the reference to the data values secret into CalicoConfig Status field
+	calicoConfig.Status.SecretRef = &cniv1alpha1.DataValuesSecretRef{
+		Name:      secretNamespacedName.Name,
+		Namespace: secretNamespacedName.Namespace,
+	}
+
+	return nil
+}
+
+func (r *CalicoConfigReconciler) ReconcileDataValuesSecret(
+	cluster *clusterapiv1beta1.Cluster,
+	calicoConfig *cniv1alpha1.CalicoConfig,
+	secretNamespacedName types.NamespacedName) error {
+
+	// prepare the data values secret for CalicoConfig
+	calicoDataValuesSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretNamespacedName.Name,
+			Namespace: secretNamespacedName.Namespace,
+		},
+		Type: corev1.SecretTypeOpaque,
+	}
+
+	calicoDataValuesSecretMutateFn := func() error {
+		calicoDataValuesSecret.Data = make(map[string][]byte)
+		clusterNetwork := cluster.Spec.ClusterNetwork
+		if clusterNetwork == nil {
+			err := errors.New("cluster.Spec.ClusterNetwork is not set")
+			r.Log.Error(err, fmt.Sprintf("Could not find ClusterNetwork in cluster '%s' Spec", cluster.Name))
+			return err
+		}
+		clusterCIDR := clusterNetwork.Pods
+		if clusterCIDR == nil || len(clusterCIDR.CIDRBlocks) == 0 {
+			clusterCIDR = clusterNetwork.Services
+		}
+		if clusterCIDR == nil || len(clusterCIDR.CIDRBlocks) == 0 {
```

Comment:
You should not look for serviceCIDR here. The ClusterCIDR is specifically `CLUSTER_CIDR` which is [pod CIDR](https://github.com/vmware-tanzu/tanzu-framework/blob/main/pkg/v1/providers/ytt/02_addons/cni/calico/calico_addon_data.lib.yaml#L14).